### PR TITLE
Add table azure_data_factory_dataset. Closes #150

### LINKED ISSUE #150: Add table azure_datafactory_dataset
**References**
https://docs.microsoft.com/en-us/azure/data-factory/quickstart-create-data-factory-rest-api

### PR DESCRIPTION
**Pull request by rajeshbal65**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_data_factory_dataset []

PRETEST: tests/azure_data_factory_dataset

TEST: tests/azure_data_factory_dataset
Running terraform
azurerm_resource_group.named_test_resource: Refreshing state... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046]
azurerm_data_factory.named_test_resource: Refreshing state... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046]
azurerm_data_factory_linked_service_mysql.named_test_resource: Refreshing state... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046/linkedservices/turbottest90046]
azurerm_data_factory_dataset_mysql.named_test_resource: Refreshing state... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046/datasets/turbottest90046]
azurerm_data_factory_dataset_mysql.named_test_resource: Destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046/datasets/turbottest90046]
azurerm_data_factory_dataset_mysql.named_test_resource: Destruction complete after 2s
azurerm_data_factory_linked_service_mysql.named_test_resource: Destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046/linkedservices/turbottest90046]
azurerm_data_factory_linked_service_mysql.named_test_resource: Destruction complete after 3s
azurerm_data_factory.named_test_resource: Destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046/providers/Microsoft.DataFactory/factories/turbottest90046]
azurerm_data_factory.named_test_resource: Destruction complete after 7s
azurerm_resource_group.named_test_resource: Destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest90046]
azurerm_resource_group.named_test_resource: Still destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-...3dae6c8/resourceGroups/turbottest90046, 10s elapsed]
azurerm_resource_group.named_test_resource: Still destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-...3dae6c8/resourceGroups/turbottest90046, 20s elapsed]
azurerm_resource_group.named_test_resource: Still destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-...3dae6c8/resourceGroups/turbottest90046, 30s elapsed]
azurerm_resource_group.named_test_resource: Still destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-...3dae6c8/resourceGroups/turbottest90046, 40s elapsed]
azurerm_resource_group.named_test_resource: Still destroying... [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-...3dae6c8/resourceGroups/turbottest90046, 50s elapsed]
azurerm_resource_group.named_test_resource: Destruction complete after 52s
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 2s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077]
azurerm_data_factory.named_test_resource: Creating...
azurerm_data_factory.named_test_resource: Still creating... [10s elapsed]
azurerm_data_factory.named_test_resource: Creation complete after 11s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077]
azurerm_data_factory_linked_service_mysql.named_test_resource: Creating...
azurerm_data_factory_linked_service_mysql.named_test_resource: Creation complete after 3s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/linkedservices/turbottest39077]
azurerm_data_factory_dataset_mysql.named_test_resource: Creating...
azurerm_data_factory_dataset_mysql.named_test_resource: Creation complete after 5s [id=/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Warning: Version constraints inside provider configuration blocks are deprecated

  on variables.tf line 22, in provider "azurerm":
  22:   version         = "=1.36.0"

Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.


Apply complete! Resources: 4 added, 0 changed, 4 destroyed.

Outputs:

resource_aka = "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077"
resource_aka_lower = "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest39077/providers/microsoft.datafactory/factories/turbottest39077/datasets/turbottest39077"
resource_id = "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077"
resource_name = "turbottest39077"
subscription_id = "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8"

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077",
    "name": "turbottest39077",
    "resource_group": "turbottest39077",
    "subscription_id": "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8",
    "type": "Microsoft.DataFactory/factories/datasets"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077",
    "name": "turbottest39077",
    "type": "Microsoft.DataFactory/factories/datasets"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest39077/providers/Microsoft.DataFactory/factories/turbottest39077/datasets/turbottest39077",
      "azure:///subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourcegroups/turbottest39077/providers/microsoft.datafactory/factories/turbottest39077/datasets/turbottest39077"
    ],
    "name": "turbottest39077",
    "title": "turbottest39077"
  }
]
✔ PASSED

POSTTEST: tests/azure_data_factory_dataset

TEARDOWN: tests/azure_data_factory_dataset

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  name,
  id,
  description,
  etag,
  type
from
  azure_data_factory_dataset;

+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------------------------------+------------------------------------------+
| name            | id                                                                                                                                                                    | description | etag                                 | type                                     |
+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------------------------------+------------------------------------------+
| turbottest99785 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99785/providers/Microsoft.DataFactory/factories/turbottest99785/datasets/turbottest99785 | <null>      | 0100cfee-0000-0100-0000-60e2cec00000 | Microsoft.DataFactory/factories/datasets |
+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------------------------------+------------------------------------------+
```

### List relational table type Datasets

```sql
select
  name,
  id,
  type,
  properties ->> 'type' as dataset_type
from
  azure_data_factory_dataset
where
  properties ->> 'type' = 'RelationalTable';

+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------+-----------------+
| name            | id                                                                                                                                                                    | type                                     | dataset_type    |
+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------+-----------------+
| turbottest99785 | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbottest99785/providers/Microsoft.DataFactory/factories/turbottest99785/datasets/turbottest99785 | Microsoft.DataFactory/factories/datasets | RelationalTable |
+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------+-----------------+
```
```
</details>

### CODE REVIEW

File: azure/table_azure_data_factory_dataset.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory"
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+)
+
+//// TABLE DEFINITION ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:14]:
```suggestion
//// TABLE DEFINITION
```

File: azure/table_azure_data_factory_dataset.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	return &plugin.Table{
+		Name:        "azure_data_factory_dataset",
+		Description: "Azure Data Factory Dataset",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.AllColumns([]string{"name", "resource_group", "factory_name"}),
+			Hydrate:           getDataset,
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFound", "ResourceGroupNotFound", "404"}),
+		},
+		List: &plugin.ListConfig{
+			Hydrate:       listDatasets,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:26]:
```suggestion
			Hydrate:       listDataFactoryDatasets,
```

File: azure/table_azure_data_factory_dataset.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+				Transform:   transform.FromField("Etag"),
+			},
+			{
+				Name:        "type",
+				Description: "The resource type.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "factory_name",
+				Description: "Time the factory was created in ISO8601 format.",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:60]:
```suggestion
				Description: "Specifies the time, the factory was created.",
```

File: azure/table_azure_data_factory_dataset.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "properties",
+				Description: "Dataset ElapsedTime Metric Policy.",
+				Type:        proto.ColumnType_JSON,
+			},
+			// Standard columns
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:68]:
```suggestion
			},
			
			// Standard columns
```

File: azure/table_azure_data_factory_dataset.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Name"),
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+			{
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:81]:
```suggestion
			},

                         // Azure standard columns
			{
```

File: azure/table_azure_data_factory_dataset.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Etag"),
+			},
+			{
+				Name:        "type",
+				Description: "The resource type.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "factory_name",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:59]: Add this column after `id`

File: azure/table_azure_data_factory_dataset.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		},
+	}
+}
+
+type DatasetInfo = struct {
+	datafactory.DatasetResource
+	FactoryName string
+}
+
+//// LIST FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:102]:
```suggestion
//// LIST FUNCTION
```

File: azure/table_azure_data_factory_dataset.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		for _, dataset := range result.Values() {
+			d.StreamListItem(ctx, DatasetInfo{dataset, *factoryInfo.Name})
+		}
+		result.NextWithContext(context.Background())
+		pagesLeft = result.NotDone()
+	}
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS ////
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:134]:
```suggestion
//// HYDRATE FUNCTIONS
```

File: azure/table_azure_data_factory_dataset.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+//// TABLE DEFINITION ////
+
+func tableAzureDataFactoryDataset(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "azure_data_factory_dataset",
+		Description: "Azure Data Factory Dataset",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.AllColumns([]string{"name", "resource_group", "factory_name"}),
+			Hydrate:           getDataset,
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:22]:
```suggestion
			Hydrate:           getDataFactoryDataset,
```

File: azure/table_azure_data_factory_dataset.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+//// HYDRATE FUNCTIONS ////
+
+func getDataset(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getDataset")
+
+	DatasetName := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+	factoryName := d.KeyColumnQuals["factory_name"].GetStringValue()
+
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:142]: Add check for empty input

File: docs/tables/azure_data_factory_dataset.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  id,
+  description,
+  etag,
+  type
+from
+  azure_data_factory_dataset;
+```
+
+
+### List relational table type Datasets
```
Comment (Subhajit97) [inline docs/tables/azure_data_factory_dataset.md:21]:
```suggestion
### List relational table type datasets
```

File: azure/table_azure_data_factory_dataset.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+			},
+			{
+				Name:        "id",
+				Description: "The resource identifier.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromGo(),
+			},
+			{
+				Name:        "factory_name",
+				Description: "Name of the factory the dataset belongs",
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:43]:
```suggestion
				Description: "Name of the factory the dataset belongs.",
```

File: azure/table_azure_data_factory_dataset.go [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "description",
+				Description: "The description of the Dataset.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Dataset.Description"),
+			},
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:51]: Is this property defined in SDK?

File: azure/table_azure_data_factory_dataset.go [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+			{
+				Name:        "etag",
+				Description: "Etag identifies change in the resource.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Etag"),
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:56]: Not required
```suggestion
```

File: docs/tables/azure_data_factory_dataset.md [hunk 15]
```diff
@@ -0,0 +1,33 @@
+# Table: azure_data_factory_dataset
+
+ Datasets identify data within different data stores, such as tables, files, folders, and documents.
```
Comment (Subhajit97) [inline docs/tables/azure_data_factory_dataset.md:3]:
```suggestion
Azure Data Factory datasets identify data within different data stores, such as tables, files, folders, and documents.
```

File: azure/table_azure_data_factory_dataset.go [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+
+			// Azure standard column
+			{
```
Comment (Subhajit97) [inline azure/table_azure_data_factory_dataset.go:84]: Add `region` column

**Review by Subhajit97 — APPROVED** [review]

**Review by rajlearner17 — APPROVED** [review]

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**